File: leetcode/lc-us/cpp/1819.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int countDifferentSubsequenceGCDs(vector<int> &nums) {
        int N = *max_element(nums.begin(), nums.end()) + 1;
        bool S[N];
        memset(S, false, sizeof S);
        for (int x: nums) S[x] = true;
        int res = 0;
        for (int x = 1; x < N; x++) {
            int gcd = -1;
            for (int i = 1; x * i < N; i++) {
                if (S[x * i]) {
                    if (gcd == -1) gcd = x * i;
                    else gcd = __gcd(gcd, x * i);
                }
            }
            if (gcd == x) res++;
        }
        return res;
    }
};

File: leetcode/lc-us/py/1819.py
from typing import List


class Solution:
    def countDifferentSubsequenceGCDs(self, nums: List[int]) -> int:
        mx = max(nums)

        def gcd(x: int, y: int) -> int:
            if x % y == 0:
                return y
            return gcd(y, x % y)

        num_set = set(nums)
        res = 0
        for i in range(1, mx + 1):
            g = -1
            for j in range(i, mx + 1, i):
                if j not in num_set:
                    continue
                if g == -1:
                    g = j
                else:
                    g = gcd(g, j)
                j += i
                if g == i:
                    res += 1
                    break
        return res
